NiL.JS/Core/Tools.cs

```diff
@@ -1887,5 +1887,245 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+        }
+
+        internal static string JSValueToObjectString(JSValue v, int maxRecursionDepth, int recursionDepth = 0)
+        {
+            if (v == null)
+                return "null";
+
+            switch (v.ValueType)
+            {
+                case JSValueType.String:
+                    return "\"" + v.ToString() + "\"";
+                case JSValueType.Date:
+                    string dstr = v.ToString();
+                    if (dstr == "Invalid date")
+                        return dstr;
+                    return "Date " + dstr;
+                case JSValueType.Function:
+                    BaseLibrary.Function f = v.Value as BaseLibrary.Function;
+                    if (recursionDepth >= maxRecursionDepth)
+                        return f.name + "()";
+                    return $"function {f.name}()";
+                case JSValueType.Object:
+                    if (v == null || v._oValue == null)
+                        return "null";
+
+                    if (v.Value is RegExp)
+                        return v.ToString();
+
+                    if (v.Value is BaseLibrary.Array
+                        || v.Value == Context.CurrentBaseContext.GetPrototype(typeof(BaseLibrary.Array)))
+                    {
+                        BaseLibrary.Array a = v.Value as BaseLibrary.Array;
+                        long len = (long)a.length;
+
+                        if (recursionDepth >= maxRecursionDepth)
+                            return $"Array[{len}]";
+
+                        StringBuilder s = new StringBuilder($"Array ({len}) [ ");
+                        int i = 0;
+                        for (i = 0; i < len; i++)
+                        {
+                            if (i > 0)
+                                s.Append(", ");
+                            s.Append(Tools.JSValueToObjectString(a[i], maxRecursionDepth, recursionDepth + 1));
+                        }
+                        if (a._fields != null)
+                        {
+                            for (var e = a._fields.GetEnumerator(); e.MoveNext();)
+                            {
+                                if (i++ > 0)
+                                    s.Append(", ");
+                                s.Append(e.Current.Key).Append(": ");
+                                s.Append(Tools.JSValueToObjectString(e.Current.Value, maxRecursionDepth, recursionDepth + 1));
+                            }
+                        }
+                        s.Append(" ]");
+
+                        return s.ToString();
+                    }
+                    else
+                    {
+                        string typeName = v.toString(null).ToString(); // "[object TYPENAME]"
+                        typeName = typeName.Substring(8, typeName.Length - 1 - 8); // 8 = "[object ".Length
+
+                        if (recursionDepth >= maxRecursionDepth)
+                            return typeName;
```

Comment:
JSValue.toString(...) not for all objects returns `[object TYPENAME]`. Maybe it would be useful to create function "GetTypeName" by extracting some code from JSValue.toString(...). With internal modifier, I think it can be modified in future.